File: ui/server/runs/plan.ts
/**
 * What a run would do, asked of the pipeline rather than worked out here.
 *
 * One subprocess — `runs/plan.py`, the adapter beside this file — one JSON document, one
 * normalisation. Which Profile a run loads, where it writes, which view it rebuilds and
 * whether this Install's Filter Decisions belong to that Profile are all questions
 * `src/venator/` already answers; the adapter composes those answers and this module reads
 * them. That is the arrangement `onboarding/boards.ts` has with `venator.discover.register`
 * and `onboarding/runtime.ts` has with `venator.llm.probe`, and it exists for the same
 * measured reason: a second copy of a pipeline rule in TypeScript drifts, and this repository
 * has already paid for one.
 *
 * Asking costs nothing. `plan.py` writes no file, creates no store, starts no stage and makes
 * no request — so unlike the runtime probe this is safe to call from a screen a person opened.
 *
 * A shape this module does not recognise is reported as a mismatch rather than guessed at,
 * on `readProbeDocument`'s own principle: naming the drift beats rendering something wrong.
 */

import { spawn } from "node:child_process";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

import type { PlanKind } from "../../shared/runs.ts";
import {
	namedPythonInterpreter,
	pipelineWorkingDirectory,
	pythonInterpreter,
	systemContext,
	type LocationContext,
} from "../locations.ts";
// The server's JSON value model. It sits under `onboarding/` because that is where it was
// first needed, not because it is an onboarding rule; nothing about reading a mapping out of
// a parsed document belongs to one surface. Imported rather than copied for the reason the
// whole of this file exists.
import { asNumber, asMapping, asText, at, parseJson } from "../onboarding/json.ts";
import { runEnvironment } from "./environment.ts";

/** The adapter that composes the pipeline's own answers. Beside this file, and only here. */
const PLAN_SCRIPT = join(fileURLToPath(new URL(".", import.meta.url)), "plan.py");

/**
 * Loading a Profile and listing a decisions directory. There is no network and no LLM in it,
 * so this bounds a machine that is thrashing rather than work that legitimately takes time.
 */
// Counting current Jev work requires replaying the local Posting and qualification stores.
const PLAN_TIMEOUT_MS = 120_000;

/** Exit 3 from the adapter: the pipeline is not importable in this Install. */
const PIPELINE_MISSING_EXIT = 3;

/** What the pipeline says a run would do. Every field is the pipeline's answer, not ours. */
export type PipelineJevPlan = {
	readonly asOf: string;
	readonly mode: "shadow" | "promoted";
	readonly postingCount: number;
	readonly maximumUsd: number;
	readonly selectionHash: string;
};

export type PipelinePlan = {
	readonly profileName: string;
	readonly profileDirectory: string;
	readonly storeRoot: string;
	readonly viewPath: string;
	readonly boards: number;
	readonly jev?: PipelineJevPlan | null;
};

/**
 * Why the pipeline will not describe a run.
 *
 * `profile` and `store` carry the pipeline's own sentence, which already names the fix.
 * `failed` carries none: the adapter classified it as unrecognised and put the original on
 * stderr, where the server logs it.
 */
export type PlanFailureKind = "profile" | "store" | "failed";

export type PlanDocument =
	| { readonly outcome: "plan"; readonly plan: PipelinePlan }
	| { readonly outcome: "failure"; readonly kind: PlanFailureKind; readonly message: string | null }
	/** The adapter answered in a shape this dashboard does not know. Never guessed at. */
	| { readonly outcome: "mismatch" }
	/** The pipeline is not part of this Install, or the interpreter would not start. */
	| { readonly outcome: "absent"; readonly message: string }
	| { readonly outcome: "timed-out" };

const PLAN_FAILURE_KINDS: readonly PlanFailureKind[] = ["profile", "store", "failed"];

/**
 * Reads the adapter's document.
 *
 * Exported for its own tests. Every field is required to be the type it is: a `boards` that
 * arrived as text, or a `viewPath` that did not arrive, is a mismatch rather than a zero or
 * an empty string, because both of those would be rendered to a person as a fact.
 */
export function readPlanDocument(text: string): PlanDocument {
	const document = asMapping(parseJson(text));
	if (document === null) return { outcome: "mismatch" };

	const failure = asMapping(at(document, "failure"));
	if (failure !== null) {
		const raw = asText(at(failure, "kind"));
		const kind = PLAN_FAILURE_KINDS.find((candidate) => candidate === raw);
		if (kind === undefined) return { outcome: "mismatch" };
		return { outcome: "failure", kind, message: asText(at(failure, "message")) };
	}

	const profile = asMapping(at(document, "profile"));
	if (profile === null) return { outcome: "mismatch" };
	const profileName = asText(at(profile, "name"));
	const profileDirectory = asText(at(profile, "directory"));
	const storeRoot = asText(at(document, "storeRoot"));
	const viewPath = asText(at(document, "viewPath"));
	const boards = asNumber(at(document, "boards"));
	const rawJev = at(document, "jev");
	const jevDocument = rawJev === undefined ? null : asMapping(rawJev);
	let jev: PipelineJevPlan | null = null;
	if (jevDocument !== null) {
		const asOf = asText(at(jevDocument, "asOf"));
		const rawMode = asText(at(jevDocument, "mode"));
		const postingCount = asNumber(at(jevDocument, "postingCount"));
		const maximumUsd = asNumber(at(jevDocument, "maximumUsd"));
		const selectionHash = asText(at(jevDocument, "selectionHash"));
		if (
			asOf === null ||
			!/^\d{4}-\d{2}-\d{2}$/u.test(asOf) ||
			(rawMode !== "shadow" && rawMode !== "promoted") ||
			postingCount === null ||
			!Number.isInteger(postingCount) ||
			postingCount < 0 ||
			maximumUsd === null ||
			!Number.isFinite(maximumUsd) ||
			maximumUsd < 0 ||
			selectionHash === null ||
			!/^[a-f0-9]{64}$/u.test(selectionHash)
		) {
			return { outcome: "mismatch" };
		}
		jev = { asOf, mode: rawMode, postingCount, maximumUsd, selectionHash };
	} else if (rawJev !== undefined) {
		return { outcome: "mismatch" };
	}
	if (
		profileName === null ||
		profileName === "" ||
		profileDirectory === null ||
		storeRoot === null ||
		viewPath === null ||
		boards === null ||
		!Number.isInteger(boards)
	) {
		return { outcome: "mismatch" };
	}
	const plan: PipelinePlan = { profileName, profileDirectory, storeRoot, viewPath, boards };
	return {
		outcome: "plan",
		plan: jev === null ? plan : { ...plan, jev },
	};
}

/**
 * What to say when the interpreter would not start at all.
 *
 * The same two facts `onboarding/runtime.ts` distinguishes, for the same reason: an Install
 * that named no interpreter and found none has no pipeline in it, while a desktop bundle that
 * shipped one and cannot start it is a broken interpreter, and telling that person the
 * pipeline is not installed sends them to fix something that is already there.
 */
function launchFailureMessage(context: LocationContext): string {
	const named = namedPythonInterpreter(context);
	if (named === null) {
		return "Running the pipeline is not part of this Install, so nothing can be started from here.";
	}
	return `The interpreter this Install runs the pipeline with could not be started, so no run can begin: ${named}`;
}

/** What the adapter's process did, as the four facts the outcome is decided from. */
export type PlanProcess = {
	readonly code: number | null;
	readonly stdout: string;
	readonly stderr: string;
	readonly timedOut: boolean;
	readonly launchFailed: boolean;
};

function runPlanScript(kind: PlanKind, profile: string, context: LocationContext): Promise<PlanProcess> {
	return new Promise((settle) => {
		const child = spawn(
			pythonInterpreter(context),
			[PLAN_SCRIPT, "--profile", profile, "--kind", kind],
			{
				cwd: pipelineWorkingDirectory(context),
				stdio: ["ignore", "pipe", "pipe"],
				env: runEnvironment(context),
			},
		);
		let stdout = "";
		let stderr = "";
		let timedOut = false;
		let launchFailed = false;
		const timer = setTimeout(() => {
			timedOut = true;
			child.kill("SIGKILL");
		}, PLAN_TIMEOUT_MS);
		child.stdout?.setEncoding("utf8");
		child.stderr?.setEncoding("utf8");
		child.stdout?.on("data", (chunk: string) => {
			stdout += chunk;
		});
		child.stderr?.on("data", (chunk: string) => {
			stderr += chunk;
		});
		child.on("error", () => {
			launchFailed = true;
		});
		child.on("close", (code) => {
			clearTimeout(timer);
			settle({ code, stdout, stderr, timedOut, launchFailed });
		});
	});
}

/**
 * Asks the pipeline what a run for `profile` would do.
 *
 * The Profile is always named. The run surface never leaves it implicit — the plan states
 * which Profile it is for and the screen shows that name, so nobody presses a button whose
 * subject they cannot see.
 */
export async function describeRun(
	kind: PlanKind,
	profile: string,
	context: LocationContext = systemContext(),
): Promise<PlanDocument> {
	return readPlanResult(await runPlanScript(kind, profile, context), context);
}

/**
 * What a finished adapter process means.
 *
 * Separated from the spawn so every branch is a pure function of four values and can be
 * asserted without a child process — which is what makes the branches testable identically on
 * every host, rather than on the one where a stand-in program happens to be startable.
 */
export function readPlanResult(finished: PlanProcess, context: LocationContext): PlanDocument {
	if (finished.launchFailed) return { outcome: "absent", message: launchFailureMessage(context) };
	if (finished.timedOut) return { outcome: "timed-out" };
	if (finished.code === PIPELINE_MISSING_EXIT) {
		return { outcome: "absent", message: launchFailureMessage(context) };
	}
	if (finished.code !== 0) {
		// Logged rather than returned: an adapter that exited non-zero prints paths and
		// environment contents, and none of that is the caller's to see.
		process.stderr.write(`runs: the plan adapter exited ${String(finished.code)}: ${finished.stderr.slice(0, 400)}\n`);
		return { outcome: "absent", message: launchFailureMessage(context) };
	}
	try {
		return readPlanDocument(finished.stdout);
	} catch {
		return { outcome: "mismatch" };
	}
}

File: ui/server/runs/plan.py
"""Report what a pipeline run would do, in JSON, for the dashboard's run surface.

    python plan.py --kind fetch-and-filter|jev-it [--profile NAME]

This is an **adapter**, not a second pipeline. Every rule it applies belongs to
``src/venator/``: which Profile a run would load and how a name resolves to a
directory (``venator.profile``), where a run would write and which view it would
rebuild (``venator.paths``), and whether this Install's Filter Decisions belong to
that Profile at all (``venator.match.store.verify_decisions_dir``). It imports
and composes; it decides nothing. A second copy of any of it in TypeScript is
the drift ``resolve_board.py`` was written to prevent, and that file's header
states the rule this one reuses verbatim.

It is the mirror of what ``server/onboarding/resolve_board.py`` does for
``venator.discover.register``: one subprocess, one JSON document on stdout, one
normalisation on the Node side.

Three properties are deliberate:

* **It writes nothing and starts nothing.** No store is created, no directory is
  stamped, no stage is run, and no request leaves this machine. Answering "what
  would a run do" must not itself be a run.
* **A refusal is data.** A Profile that cannot be chosen, or a store this
  Profile may not read, comes back as a ``failure`` object with the pipeline's
  own sentence in it — exit 0 — so the dashboard can put that sentence next to a
  disabled button rather than rendering a crash.
* **A caught exception's text is never invented over.** ``ProfileError`` and the
  store-claim refusal are written for a person by the modules that raise them,
  and they are passed along unchanged. Everything else is a category with a
  message written here, and the original goes to stderr where the server logs it.

Exit codes: ``0`` a JSON document on stdout, ``3`` the pipeline module could not
be imported, ``2`` this script was called wrongly.
"""

from __future__ import annotations

import argparse
import contextlib
import hashlib
import json
import sys
from datetime import date
from pathlib import Path

# The pipeline prints its Profile and store announcements as it resolves them.
# They go to stderr already, but `resolve_profile` and friends may print to
# stdout too, so stdout is held aside and handed back only for the document.
_STDOUT = sys.stdout

try:
    from venator.match.store import verify_decisions_dir
    from venator.paths import StoreRootError, install_stores
    from venator.profile import ProfileError, resolve_profile
    from venator.qualify.jev import current_mode, passing_postings, select_work
    from venator.qualify.jev_release import JEV_RELEASE
    from venator.qualify.store import as_of_month
except Exception as err:  # pragma: no cover - exercised by an Install without the pipeline
    print(f"plan: the pipeline could not be imported: {err}", file=sys.stderr)
    raise SystemExit(3) from err


def _failure(kind: str, message: str | None) -> dict:
    return {"failure": {"kind": kind, "message": message}}


def _jev_plan(profile, stores, as_of: str) -> dict:
    month = as_of_month(as_of)
    mode = current_mode(profile, month, stores.qualifications_dir, JEV_RELEASE)
    postings = passing_postings(
        stores.postings_dir, stores.decisions_dir, stores.qualifications_dir, profile, month,
    )
    work = select_work(
        postings,
        profile=profile,
        month=month,
        mode=mode,
        qualifications_dir=stores.qualifications_dir,
        named_keys=None,
        release=JEV_RELEASE,
        decided_at=f"{as_of}T00:00:00+00:00",
        bindings_only=True,
    )
    count = work.report.selected
    # The count and rounded allowance alone do not identify the work. A Profile
    # edit or a different set of Postings can leave both unchanged.
    selection_hash = hashlib.sha256(json.dumps(sorted(
        work.selection
    ), separators=(",", ":")).encode("utf-8")).hexdigest()
    maximum_usd = 10.0 if count else 0.0
    return {
        "asOf": as_of,
        "mode": mode,
        "postingCount": count,
        "maximumUsd": maximum_usd,
        "selectionHash": selection_hash,
    }


def _document(profile, stores, as_of: str, kind: str) -> dict:
    document = {
        "profile": {
            "name": profile.name,
            # Absolute, because the Node side compares it against the Profile
            # the dashboard listed and shows it to a person; the pipeline's own
            # `profiles/<name>` is relative to whatever it was run from.
            "directory": str(Path(profile.directory).resolve()),
            "identifier": profile.identifier,
        },
        "storeRoot": str(stores.path),
        "viewPath": str(stores.database_path),
        # How many ATS boards this Profile registers. Zero is not an error and
        # is not a guess about the search: it is why a first run can finish
        # cleanly and find nothing, which is worth saying before the run rather
        # than after it.
        "boards": sum(len(tokens) for tokens in profile.sources.boards.values()),
    }
    if kind == "jev-it":
        document["jev"] = _jev_plan(profile, stores, as_of)
    return document


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(description="What a pipeline run would do.")
    parser.add_argument("--profile", default=None)
    parser.add_argument("--kind", choices=("fetch-and-filter", "jev-it"), required=True)
    args = parser.parse_args(argv)

    with contextlib.redirect_stdout(sys.stderr):
        try:
            profile = resolve_profile(requested=args.profile)
        except ProfileError as err:
            # `venator.profile` writes these for the person reading them: no
            # Profile, two Profiles and which, or a name that is not one of
            # them. Passed through unchanged.
            payload = _failure("profile", str(err))
        else:
            try:
                stores = install_stores()
            except StoreRootError as err:
                # Already names both candidate stores and the fix.
                payload = _failure("store", str(err))
            else:
                try:
                    # The read that a run would do first. A decisions store
                    # claimed by another Profile is refused here, before
                    # Discover has appended anything, rather than partway
                    # through a run — `data/` is append-only.
                    if stores.decisions_dir.exists():
                        verify_decisions_dir(stores.decisions_dir, profile.identifier)
                    payload = _document(profile, stores, date.today().isoformat(), args.kind)
                except (OSError, ValueError) as err:
                    payload = _failure("store", str(err))
                except Exception as err:  # pragma: no cover - the unclassified case
                    print(f"plan: {type(err).__name__}: {err}", file=sys.stderr)
                    payload = _failure("failed", None)

    json.dump(payload, _STDOUT)
    _STDOUT.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
